# Grouping Data and Aggregating Variables

## Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
# Assigning the path

path = r'C:\Users\guest23455\Desktop\CareerFoundry\Data Immersion\Achievement 4_Python Fundamentals for Data Analysts\05-2023 Instacart Basket Analysis'

In [3]:
# Import orders_products_merged_updated

ords_prods_merge =  pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','orders_products_merged_updated.pkl'))

In [4]:
# Create a subset with first 1 million rows

df = ords_prods_merge[:1000000]

In [5]:
# How many rows and columns
df.shape

(1000000, 19)

In [6]:
# Display the first 10 rows

df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   
5                   19.0         196                  1          1   
6                   20.0         196                  1          1   
7                   14.0         196                  2          1   
8                    0.0         196                  4          1   
9                   30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0         Soda        77              7     9.0   both              NaN   
1         Soda        77              7     9.0   both              NaN   
2         Soda        77              7     9.0   both              NaN   
3         Soda        77              7     9.0   both              NaN   
4         Soda        77              7     9.0   both              NaN   
5         Soda        77              7     9.0   both              NaN   
6         Soda        77              7     9.0   both              NaN   
7         Soda        77              7     9.0   both              NaN   
8         Soda        77              7     9.0   both              NaN   
9         Soda        77              7     9.0   both              NaN   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy        Average orders  
1  Mid-range product  Regularly busy      Least busy        Average orders  
2  Mid-range product  Regularly busy      Least busy           Most orders  
3  Mid-range product      Least busy      Least busy        Average orders  
4  Mid-range product      Least busy      Least busy           Most orders  
5  Mid-range product  Regularly busy  Regularly busy        Average orders  
6  Mid-range product  Regularly busy     Busiest day           Most orders  
7  Mid-range product  Regularly busy     Busiest day           Most orders  
8  Mid-range product  Regularly busy     Busiest day           Most orders  
9  Mid-range product      Least busy      Least busy        Average orders

## Grouping Data with Pandas

In [7]:
# Let us group the df by product_name using the groupby() 

df.groupby('product_name')

# Step 1 : Splitting the data into groups based on some criteria

## Aggregating Data

### Aggregating Data with agg()

In [8]:
# Step 2 : Apply a function to each group separately

df.groupby('department_id').agg({'order_number': ['mean']})

# In the above code the first two steps are done together
# The df is grouped by the 'department_id' (Step 1)
# Then the agg() function is applied to the group, to obtain the mean values for 'order_number' (Step 2)

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

It displays the average number of orders for each department

In [9]:
# The same can be done without the use of agg()

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# Performing multiple aggregations

df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

# Aggregating the group by mean, min, and max values of 'order_number'

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### Aggregating Data with transform()

In [11]:
# Create a loyalty flag using loc

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

# Step 1 : Split the data into groups based on the 'user_id' column
# Step 2 : Apply the transform() function to the 'order_number' column to generate te maximum orders for each user
# Step 3 : Create a new column 'max_order', into which the results of the aggregation will be placed

In [15]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0          Soda        77              7     9.0   both              NaN   
1          Soda        77              7     9.0   both              NaN   
2          Soda        77              7     9.0   both              NaN   
3          Soda        77              7     9.0   both              NaN   
4          Soda        77              7     9.0   both              NaN   
5          Soda        77              7     9.0   both              NaN   
6          Soda        77              7     9.0   both              NaN   
7          Soda        77              7     9.0   both              NaN   
8          Soda        77              7     9.0   both              NaN   
9          Soda        77              7     9.0   both              NaN   
10         Soda        77              7     9.0   both              NaN   
11         Soda        77              7     9.0   both              NaN   
12         Soda        77              7     9.0   both              NaN   
13         Soda        77              7     9.0   both              NaN   
14         Soda        77              7     9.0   both              NaN   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy        Average orders   
1   Mid-range product  Regularly busy      Least busy        Average orders   
2   Mid-range product  Regularly busy      Least busy           Most orders   
3   Mid-range product      Least busy      Least busy        Average orders   
4   Mi

## Deriving Columns with loc()

In [16]:
# Set the loyalty flag

# If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer”

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
# If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer”

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [18]:
# If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer”

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
# Check value counts

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [22]:
# Verify 

ords_prods_merge[['user_id', 'order_number', 'loyalty_flag']].head(15)

user_id  order_number      loyalty_flag
0         1             1      New customer
1         1             2      New customer
2         1             3      New customer
3         1             4      New customer
4         1             5      New customer
5         1             6      New customer
6         1             7      New customer
7         1             8      New customer
8         1             9      New customer
9         1            10      New customer
10       15            15  Regular customer
11       15            17  Regular customer
12       15            18  Regular customer
13       15            21  Regular customer
14       15            22  Regular customer

**Exercise 4.8**

# Task 2 

In [24]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Task 3

1. First difference is that we can see all the departments, where as in the sample data we had only few of the departments.
2. The mean order values for the subset is larger than that of the entire dataframe, except for department_id 17.

# Task 4 : Loyal Customers
For task 4 please see above (Section 1.4.2 and 1.5)

# Task 5

In [26]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

It can be clearly seen that although New customers have less than or equal to 10 orders, the average price is 13.3$ which is more than the Loyal customers. 
Although the Loyal customers have more than 40 orders, the average price spent is less.

# Task 6 : Type of Spenders

In [27]:
# Create a spent_flag based on the average prices

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [28]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  prices  _merge   price_range_loc    price_range_loc  \
0         Soda  ...     9.0    both               NaN  Mid-range product   
1         Soda  ...     9.0    both               NaN  Mid-range product   
2         Soda  ...     9.0    both               NaN  Mid-range product   
3         Soda  ...     9.0    both               NaN  Mid-range product   
4         Soda  ...     9.0    both               NaN  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy      Least busy        Average orders        10   
2  Regularly busy      Least busy           Most orders        10   
3      Least busy      Least busy        Average orders        10   
4      Least busy      Least busy           Most orders        10   

   loyalty_flag  avg_price  
0  New customer   6.367797  
1  New customer   6.367797  
2  New customer   6.367797  
3  New customer   6.367797  
4  New customer   6.367797  

[5 rows x 22 columns]

In [29]:
# Set the spent_flag

# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”

ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spent_flag'] = 'Low spender'

In [30]:
# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender”

ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spent_flag'] = 'High spender'

In [31]:
# Check value counts

ords_prods_merge['spent_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spent_flag, dtype: int64

In [42]:
# Verify 

ords_prods_merge[['user_id', 'avg_price', 'spent_flag']].head(100)

user_id  avg_price    spent_flag
0         1   6.367797   Low spender
1         1   6.367797   Low spender
2         1   6.367797   Low spender
3         1   6.367797   Low spender
4         1   6.367797   Low spender
..      ...        ...           ...
95      360  10.006250  High spender
96      377   8.496552   Low spender
97      387   7.396610   Low spender
98      420   7.387805   Low spender
99      420   7.387805   Low spender

[100 rows x 3 columns]

# Task 7 : Frequent vs Non-Frequent Customers

In [43]:
# Create a frequency_flag

ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [44]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...   price_range_loc    price_range_loc     busiest_day  \
0         Soda  ...               NaN  Mid-range product  Regularly busy   
1         Soda  ...               NaN  Mid-range product  Regularly busy   
2         Soda  ...               NaN  Mid-range product  Regularly busy   
3         Soda  ...               NaN  Mid-range product      Least busy   
4         Soda  ...               NaN  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag avg_price  \
0  Regularly busy        Average orders        10  New customer  6.367797   
1      Least busy        Average orders        10  New customer  6.367797   
2      Least busy           Most orders        10  New customer  6.367797   
3      Least busy        Average orders        10  New customer  6.367797   
4      Least busy           Most orders        10  New customer  6.367797   

    spent_flag  median_days  
0  Low spender         20.0  
1  Low spender         20.0  
2  Low spender         20.0  
3  Low spender         20.0  
4  Low spender         20.0  

[5 rows x 24 columns]

In [45]:
# Set the frequency_flag

# If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer”

ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [46]:
# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer” 

ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [47]:
# If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer”

ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [48]:
# Check value counts

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: frequency_flag, dtype: int64

# Task 8 : Export Dataframe as a Pickle File

In [49]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_aggregated.pkl'))